### PR TITLE
[c++] Fix sanitizer flags and move from an option to a config type

### PR DESCRIPTION
**Pull request by jp-dark**
The builds using the `SANITIZER` option were missing the linker flags necessary to build. This PR adds those back in and moves enabling sanitizers to a config option. The new build types are enabled:

* `ASAN`: address sanitizer 
* `TSAN`: thread sanitizer
* `LSAN`: leak sanitizer
* `MSAN`: memory sanitizer
* `UBSAN`: undefined-behavior sanitizer

### CODE REVIEW

**Review by johnkerl — COMMENTED** [review]
LGTM as long as the sanitizer is sanitizing

I defer to @nguyenv here ...

**Review by johnkerl — APPROVED** [review]
LGTM as long as the sanitizer is sanitizing

I defer to @nguyenv here ...

**Review by nguyenv — APPROVED** [review]

File: libtiledbsoma/cmake/Superbuild.cmake [hunk 1]
```diff
@@ -63,8 +63,8 @@ set(INHERITED_CMAKE_ARGS
   -DTILEDB_SERIALIZATION=${TILEDB_SERIALIZATION}
   -DTILEDB_WERROR=${TILEDB_WERROR}
   -DTILEDB_VERBOSE=${TILEDB_VERBOSE}
+  -DTILEDB_SANITIZER=${TILEDB_SANITIZER}
```
Comment (teo-tsirpanis) [inline libtiledbsoma/cmake/Superbuild.cmake:66]:
```suggestion
```

We don't need this with the multi-config arrangement, right?

File: libtiledbsoma/CMakeLists.txt [hunk 2]
```diff
@@ -83,9 +84,19 @@ set(CMAKE_CXX_EXTENSIONS OFF) # Don't use GNU extensions
...
-  set(CMAKE_BUILD_TYPE RELEASE)
+# Set default builds/configuration to be Release.
+get_property(is_multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if (is_multi_config)
+  set(CMAKE_CONFIGURATION_TYPES
+      "Release;Debug;RelWithDebInfo"
+      CACHE
+      STRING
+      "Semi-colon separate list of build types for multi-configuration generators."
+  )
```
Comment (teo-tsirpanis) [inline libtiledbsoma/CMakeLists.txt:95]:
```suggestion
  set(CMAKE_CONFIGURATION_TYPES
      "Release;Debug;RelWithDebInfo;ASAN;TSAN;LSAN;UBSAN;MSAN"
      CACHE
      STRING
      "Semi-colon separate list of build types for multi-configuration generators."
  )
  set(CMAKE_MAP_IMPORTED_CONFIG_ASAN Debug)
  set(CMAKE_MAP_IMPORTED_CONFIG_TSAN Debug)
  set(CMAKE_MAP_IMPORTED_CONFIG_LSAN Debug)
  set(CMAKE_MAP_IMPORTED_CONFIG_UBSAN Debug)
  set(CMAKE_MAP_IMPORTED_CONFIG_MSAN Debug)
```

File: libtiledbsoma/CMakeLists.txt [hunk 3]
```diff
@@ -66,6 +66,7 @@ option(TILEDB_VERBOSE "If true, sets default logging to verbose for TileDB" OFF)
 option(OVERRIDE_INSTALL_PREFIX "Ignores the setting of CMAKE_INSTALL_PREFIX and sets a default prefix" ON)
 option(ENABLE_ARROW_EXPORT "Installs an extra header for exporting in-memory results with Apache Arrow" ON)
 option(TILEDB_LOG_OUTPUT_ON_FAILURE "If true, print error logs if dependency sub-project build fails" ON)
+option(TILEDB_SANITIZER "Sanitizer to use in TILEDB. ")
```
Comment (teo-tsirpanis) [inline libtiledbsoma/CMakeLists.txt:69]:
```suggestion
```

File: libtiledbsoma/src/CMakeLists.txt [hunk 4]
```diff
@@ -251,20 +253,14 @@ if(TILEDBSOMA_BUILD_CLI)
...
-    # CLI11::CLI11
-    # spdlog::spdlog
-    tiledbsoma
-    TileDB::tiledb_shared
+      # CLI11::CLI11
+      # spdlog::spdlog
+      tiledbsoma
+      TileDB::tiledb_shared
+    PRIVATE
+      ${TILEDBSOMA_SANITIZER_FLAG}
```
Comment (teo-tsirpanis) [inline libtiledbsoma/src/CMakeLists.txt:261]: This should better go to `target_link_options`, but `libraries` works as well.

**Review by teo-tsirpanis — COMMENTED** [review]
It might have been cleaner if the sanitizer was an option separate from the config, and either accept that it will work only in Debug mode, or enforce it by setting `CMAKE_CONFIGURATION_TYPES` to only `Debug`.

File: libtiledbsoma/cmake/Superbuild.cmake (hunk 1 above)
Comment (jp-dark) [reply libtiledbsoma/cmake/Superbuild.cmake:66]: Enabling the sanitizer on TileDB embedded is a separate option.

Comment (teo-tsirpanis) [reply libtiledbsoma/cmake/Superbuild.cmake:66]: Oh I forgot about that.

File: libtiledbsoma/CMakeLists.txt (hunk 2 above)
Comment (jp-dark) [reply libtiledbsoma/CMakeLists.txt]: Thanks!

**Comment by jp-dark** [comment]
[sc-53780](https://app.shortcut.com/tiledb-inc/story/53780/fix-address-sanitizer-in-libtiledbsoma-cmake)

**Pull request merged**